quickwit/quickwit-indexing/src/actors/doc_processor.rs

```diff
@@ -68,169 +103,175 @@ impl InputDoc {
                 map.insert(key, value);
                 VrlValue::Object(map)
             }
+            InputDoc::OtlpTraceJson(_) | InputDoc::OtlpTraceProtobuf(_) => {
+                panic!("OTP log or trace data does not support VRL transforms")
+            }
         };
         Ok(vrl_doc)
     }
 
-    fn try_into_json_doc(self) -> Result<JsonObject, DocProcessorError> {
-        let json_doc = match self {
-            InputDoc::Json(doc_bytes) => serde_json::from_slice::<JsonObject>(&doc_bytes)?,
+    fn try_into_json_docs(self) -> JsonDocIterator {
+        match self {
+            InputDoc::Json(doc_bytes) => match serde_json::from_slice::<JsonObject>(&doc_bytes) {
+                Ok(json_obj) => JsonDocIterator::from(json_obj),
+                Err(error) => JsonDocIterator::from(error),
+            },
+            InputDoc::OtlpTraceJson(doc_bytes) => match parse_otlp_spans_json(&doc_bytes) {
+                Ok(spans) => JsonDocIterator::Spans(spans),
+                Err(error) => JsonDocIterator::from(error),
+            },
+            InputDoc::OtlpTraceProtobuf(doc_bytes) => match parse_otlp_spans_protobuf(&doc_bytes) {
+                Ok(spans) => JsonDocIterator::Spans(spans),
+                Err(error) => JsonDocIterator::from(error),
+            },
             InputDoc::PlainText(doc_bytes) => {
                 let mut map = serde_json::Map::with_capacity(1);
                 let key = PLAIN_TEXT.to_string();
-                let value = String::from_utf8(doc_bytes.to_vec())?;
+                let value = match String::from_utf8(doc_bytes.to_vec()) {
+                    Ok(value) => value,
+                    Err(error) => return JsonDocIterator::from(error),
+                };
                 map.insert(key, JsonValue::String(value));
-                map
+                JsonDocIterator::from(map)
             }
-        };
-        Ok(json_doc)
+        }
     }
 }
 
-#[derive(Debug)]
-pub enum DocProcessorError {
-    ParsingError,
-    MissingField,
-    #[cfg(feature = "vrl")]
-    TransformError(VrlTerminate),
+enum JsonDocIterator {
```

Comment:
This is a bit cumbersome. Do you have any suggestions for making this better? Allocating...?

Reply:
A quick and dirty alternative with similar perf is to have a `SmallVec<Result<JsonObject, DocProcessorError>, 1>`